journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,36 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.effect.implicits.effectResourceOps
+import cats.effect.std.Queue
+import cats.effect.{Concurrent, Resource}
+import cats.syntax.all._
+
+private[journal] trait ResourcePool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object ResourcePool {
+
+  /**
+   *  Objects are allocated when a pool is created and deallocated when it shuts down.
+   */
+  def fixedSize[F[_]: Concurrent, O](makeObject: Resource[F, O], poolSize: Int): Resource[F, ResourcePool[F, O]] = {
+    for {
+      objects <- Vector.fill(poolSize)(makeObject).sequence
```

Comment:
should it be `lazy`? do we want to keep the pool `full` even if we never need it?

Reply:
I've written a lazy pool